clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -217,7 +242,7 @@ private boolean isTransitionValid(State source, State target) {
                 case UNINITIALIZED:
                     return source == READY;
                 case INITIALIZING:
-                    return source == UNINITIALIZED;
+                    return source == UNINITIALIZED || source == ABORTING_TRANSACTION;
```

Comment:
We generally transition when we enqueue the relevant request, and we enqueue the abort and init at the same time. To my mind that means we go straight from `ABORTING` to `INITIALIZING`, but I'm not wedded to it if you think going through `UNINITIALIZED` makes more sense.

Reply:
I don't feel strongly about it, I was considering it mostly for consistency. For the idempotent producer, we could have transitioned directly from READY to INITIALIZING, but we chose to go through UNINITIALIZED because we thought it simplified the state machine. If we allow the transition through ABORTING, perhaps we should allow it through READY as well. Or let them both go through UNINITIALIZED.

Reply:
In that case, I think I would prefer to remove UNINITIALIZED from that path and go directly from READY to INITIALIZING. Removing the producer ID should be enough signal in that case that we need to get a new producer ID.

Reply:
Hmm, blocking READY -> INITIALIZING is how we catch duplicate `initializeTransactions` calls, so keeping that is actually useful. And going from ABORTING to UNINITIALIZED would actually open up a window where a new `initializeTransactions` call would not be blocked, which seems bad. I think I'd prefer to keep this as is, anything else actually seems more complicated. It makes the transactional and idempotent paths slightly different, but I think it's more indicative of what each path is doing (we never set the ID and epoch to `None` in the transactional path, for example).